---
title: "Analysis Notes"
author: "Leann M. Biancani"
output: html_document
---
Scripts and analyses associated with the Placental Mammal Polytomy Project. \
GitHub Repository: [PlacentalPolytomy](https://github.com/LMBiancani/PlacentalPolytomy)

Scripts adapted from [AlexKnyshov/TreeshrewProject](https://github.com/AlexKnyshov/TreeshrewProject).

```{css backgrounds, echo=FALSE}
/* Define a background class for displaying content of slurm submission text files */
.bg_slurm {
  background-color: aliceblue;
  border: 1px solid black;
}
/* Define a background class for displaying content of python text files */
.bg_python {
  background-color: lightyellow;
  border: 1px solid black;
}
/* Define a background class for displaying content of text files */
.bg_text {
  background-color: honeydew;
  border: 1px solid black;
}
/* Define a background class for displaying content of text files */
.bg_Rcode {
  background-color: lightsteelblue;
  border: 1px solid black;
}
```
***
# Data Processing
<!-- and Data Selection? -->
***
## 00. Orthology assessment using SISRS v2.0
Scripts are located in the [00_SISRS](https://github.com/LMBiancani/PlacentalPolytomy/tree/main/00_SISRS) folder.
Scripts adapted from [SchwartzLabURI/SISRS](https://github.com/SchwartzLabURI/SISRS)


***
## Phylogenetic Signal


## 01. Filter By Taxa
Scripts are located in the [01_FilterByTaxa](https://github.com/LMBiancani/PlacentalPolytomy/tree/main/01_FilterByTaxa) folder.

* Removes excessively incomplete sequences (ex. sequence must be less than 33% Ns)
* Removes aligned loci with too few taxa represented (ex. locus must include sequences for at least 25 taxa)
* Removes aligned loci with incomplete clade sampling (ex. locus must include a sequence from all 4 taxon groups included in the `groups.csv` file)
* The required taxon-to-group correspondence table, `groups.csv`, is a csv file with the following format: 
```
Group,Taxa
group1,taxonName1
group1,taxonName2
group2,taxonName3
group2,taxonName4
```
#### Slurm submission script: `01.00_filter_SISRS_output.sh`
* runs the following python script: `filter_SISRS_output.py`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 01_FilterByTaxa/01.00_filter_SISRS_output.sh
```
#### Python script: `filter_SISRS_output.py`
*  run by previous shell script: `01.00_filter_SISRS_output.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat 01_FilterByTaxa/filter_SISRS_output.py
```
#### Taxon-to-group table: `groups.csv`
* specify path to this input file in previous shell script: `01.00_filter_SISRS_output.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat 01_FilterByTaxa/groups.csv
```

## 2. Annotate Loci
Scripts are located in the [annotation](https://github.com/LMBiancani/PlacentalPolytomy/tree/main/annotation) folder.

* Aligns SISRS loci of a particular taxon to a reference genome (ideally, of the same taxon).
* A custom script is used to retreive a reference taxon for the loci.
* BLAST is run. 
* A custom python script is used to filter the output and convert it to BED. Overlapping hits of similar scores as well as very disjunct alignments are discarded. The BED file is then sorted and intersected with the GFF file for the reference sequence.
* A custom python script then processess the intersected BED file to produce the final output,
either counts of different annotation types, or length proportion of each annotations type. The following types of annotations are recorded: pseudogene, CDS, UTR, intron, lnc_RNA, other (any other type), unannotated (or intergenic).

#### Slurm submission script: `download_reference.sh`
* downloads reference genome (FASTA) and annotation (GFF) for Pan troglodytes (chimpanzee) [GCF_002880755.1](https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_002880755.1/)
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat annotation/download_reference.sh
```
#### Slurm submission script: `annotation_job.sh`
* runs the following python scripts: `annotation_getTaxContigs.py`, `annotation_blast_parser.py`, `annotation_bed2table.py`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat annotation/annotation_job.sh
```
#### Python script: `annotation_getTaxContigs.py`
* run by previous shell script: `annotation_job.sh`
* extracts SISRS loci of a particular taxon into a single file for use with BLAST
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat annotation/annotation_getTaxContigs.py
```
#### Python script: `annotation_blast_parser.py`
* BLAST to BED python script run by the previous shell script: `annotation_job.sh`
* adjust parameters on lines 10-12 as needed
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat annotation/annotation_blast_parser.py
```
#### Python script: `annotation_bed2table.py`
* BED to table python script run by the previous shell script: `annotation_job.sh`
* two options are provided (to be specified in shell script: `annotation_job.sh`)
1. count the number of each feature per locus (`c`), for ex. 1 CDS, 2 introns, etc.
2. compute proportion of length of each feature type per locus (`l`), for ex. 0.2 CDS, 0.8 introns
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat annotation/annotation_bed2table.py
```

## 3. Assess Locus Properties
### 3.1. AMAS
Scripts are located in the [amas](https://github.com/LMBiancani/PlacentalPolytomy/tree/main/amas) folder.

* Runs AMAS to assess locus features
* **Note:** AMAS takes file names as command line arguments with a limit on how long the line can be. In order to analyze several hundred thousand files, AMAS is run in batches by the driver script (`run_amas.py`).

#### Download AMAS
* note path to `AMAS.py` and add path to slurm script `run_amas.sh`
```
git clone https://github.com/marekborowiec/AMAS/
```
#### Slurm submission script: `run_amas.sh`
* runs the following python script: `run_amas.py`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat amas/run_amas.sh
```
#### Python script: `run_amas.py`
* python script run by shell script: `run_amas.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat amas/run_amas.py
```

## 4. Asses Phylogenetic Signal: IQ-TREE
Scripts are located in the [iqtree](https://github.com/LMBiancani/PlacentalPolytomy/tree/main/iqtree) folder.

#### Slurm submission script: `iqtree_prep.sh`
* sets up the folders and lists of files to process
* generates batch fasta files for aligned loci
* generates the array details that need to be added to subsequent array submission scripts
* Note: array details generated by my analysis: `#SBATCH --array=[1-28]%28`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat iqtree/iqtree_prep.sh
```

#### Slurm submission script: `iqtree_array.sh`
* runs the following R scripts: `trimTrees.R`, `getSCF.R`
* update paths to specified input files (described below)
* generates CSV files (one for each batch fasta file created by prep script) in the `$array_work_dir`
* CSV files contain likelihoods scores for each hypothesis (3 scores) for every Locus
* generates SCF files (one fore each contig) in `scf` folder
* SCF files have 3 lines (one for each hypothesesis)
* Note: update the header with the array details generated by `iqtree_prep.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat iqtree/iqtree_array.sh
```

#### Topology Constraints
Collapsed trees were created for each alternative hypothesis and used as constraint trees in IQtree

Scaffold hypotheis trees:
```
(((XENARTHRA,BOREOEUTHERIA),AFROTHERIA),METATHERIA);
(((AFROTHERIA,XENARTHRA),BOREOEUTHERIA),METATHERIA);
(((AFROTHERIA,BOREOEUTHERIA),XENARTHRA),METATHERIA);
```
Individual collapsed nodes:

METATHERIA:
```
(Didelphis_virginiana,Thylacinus_cynocephalus,Sarcophilus_harrisii,Vombatus_ursinus,Phascolarctos_cinereus,Phalanger_gymnotis,Gymnobelideus_leadbeateri,Pseudochirops_corinnae,Wallabia_bicolor,Potorous_gilbertii)
```
AFROTHERIA:
```
(Procavia_capensis,Loxodonta_africana,Hydrodamalis_gigas,Trichechus_manatus_latirostris,Orycteropus_afer,Elephantulus_edwardii,Amblysomus_hottentotus_longiceps,Chrysochloris_asiatica,Microgale_talazaci,Echinops_telfairi)
```
BOREOEUTHERIA:
```
(Condylura_cristata,Ceratotherium_simum,Manis_javanica,Odobenus_rosmarus_divergens,Hippopotamus_amphibius,Pteropus_vampyrus,Galeopterus_variegatus,Pan_troglodytes,Rattus_norvegicus,Tupaia_tana)
```
XENARTHRA:
```
(Mylodon_darwinii,Choloepus_didactylus,Myrmecophaga_tridactyla,Tamandua_tetradactyla,Dasypus_novemcinctus,Tolypeutes_matacus,Chaetophractus_vellerosus)
```

#### Alternative Hypothesis Trees: `Polytomy_Placental_Hypotheses.tree`
* text file containing alternative hypotheses trees (in Newick format)
* specify path to this input file in previous shell script: `iqtree_array.sh`
* Keep note of tree order in file:
1. Afrotheria out (Boreoeutheria and Xenarthra are sisters)
2. Boreoeutheria out (Afrotheria and Xenarthra are sisters)
3. Xenarthra out (Afrotheria and Boreoeutheria are sisters)
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat iqtree/hypothesis_trees/Polytomy_Placental_Hypotheses.tree
```

### Specify a list of focal taxa for each hypothesis tree:
* For placental root topology question: determine which 2 out of 3 groups are sisters for each hypothesis and select one of these 2 sister groups.

#### Focal taxa for tree 1: `tips_Xenarthra.txt`
* Tree 1 = Afrotheria Out, focal taxa = Xenarthra
* specify path to this input file in previous shell script: `iqtree_array.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat iqtree/hypothesis_trees/tips_Xenarthra.txt
```

#### Focal taxa for tree 2: `tips_Afrotheria.txt`
* Tree 2 = Boreoeutheria Out, focal taxa = Afrotheria
* specify path to this input file in previous shell script: `iqtree_array.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat iqtree/hypothesis_trees/tips_Afrotheria.txt
```

#### Focal taxa for tree 3: `tips_Boreoeutheria.txt`
* Tree 3 = Xenarthra Out, focal taxa = Boreoeutheria
* specify path to this input file in previous shell script: `iqtree_array.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat iqtree/hypothesis_trees/tips_Boreoeutheria.txt
```

#### Outgroup taxa list: `tips_Outgroup.txt`
* specify path to this input file in previous shell script: `iqtree_array.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat iqtree/hypothesis_trees/tips_Outgroup.txt
```

#### R script: `trimTrees.R`
* run by the following shell scripts: `iqtree_array.sh` `iqtree_array_concat.sh`
* trim taxa off of the main tree(s) if not included in taxon composition of a particular alignment
```{bash, comment=NA, echo=FALSE, class.output="bg_Rcode"}
cat iqtree/trimTrees.R
```

#### R script: `getSCF.R`
* run by previous shell script: `iqtree_array.sh`
* extract sCF values for the branch of interest from IQ-TREE output
```{bash, comment=NA, echo=FALSE, class.output="bg_Rcode"}
cat iqtree/getSCF.R
```


#### Slurm submission script: `iqtree_array_concat.sh`
* runs AMAS to infer concatenated trees and iqtree to calculate likelihoods
* runs the following R scripts: `trimTrees.R`
* Note: update paths to specified input files
* Note: update the header with the array details generated by `iqtree_prep.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat iqtree/iqtree_array_concat.sh
```


* iqtree_array_gtree.sh - slurm script to submit: infer gene trees
* iqtree_collect_gtrees.sh - slurm script to submit: collect gene tree data
* iqtree_collect_output.sh - slurm script to submit: collect individual fit assessment data
* iqtree_collect_phyloinference_LnL.sh - slurm script to submit: collect concatenation fit assessment data
* trimTrees.R - script to trim taxa off of the main tree(s) depending on the taxon composition of a particular alignment, run by the previous shell script.
* getSCF.R - script to extract sCF values for the branch of interest from IQ-TREE output, run by the previous shell script.

***