---
title: "Analysis Notes"
author: "Leann M. Biancani"
output: html_document
---
Scripts and analyses associated with the Placental Mammal Polytomy Project. \
GitHub Repository: [PlacentalPolytomy](https://github.com/LMBiancani/PlacentalPolytomy)

Some scripts adapted from [AlexKnyshov/TreeshrewProject](https://github.com/AlexKnyshov/TreeshrewProject) and [AlexandraWalling/PML](https://github.com/alexandrawalling/PML).

```{css backgrounds, echo=FALSE}
/* Define a background class for displaying content of slurm submission text files */
.bg_slurm {
  background-color: aliceblue;
  border: 1px solid black;
}
/* Define a background class for displaying content of python text files */
.bg_python {
  background-color: lightyellow;
  border: 1px solid black;
}
/* Define a background class for displaying content of text files */
.bg_text {
  background-color: honeydew;
  border: 1px solid black;
}
/* Define a background class for displaying content of text files */
.bg_Rcode {
  background-color: lightsteelblue;
  border: 1px solid black;
}
```

***
# 00. Orthology assessment using SISRS v2.0
The [SISRS v2.0](https://github.com/SchwartzLabURI/SISRS/releases/tag/v2.0) pipeline is used to obtain putatively orthologous loci from raw genomic sequencing reads.\
Required SISRS steps: 1, 2, 3, 4, 5, 6, 6b, 6c, 6d, 7, 7a, and 7b.

* In SISRS step 3 the approximate genome size was specified as 3,500,000,000 bases. (```GENOME=3500000000```)
* In SISRS step 6d the number of reads required to call a site (```MINREAD=3```) and the proportion of sites required to be the same to call a site (```THRESHOLD=1```) were left as defaults.
* In SISRS step 7 the number of species allowed to be missing was set to 25. (```-m 25```)
<!-- * In the runRAxML step specify the path to the pi_m25_nogap.phylip-relaxed alignment generated by step 7. -->
* In SISRS step 7a, the taxa threshold parameter was set to 4 (```T=4```) and the coverage and heterozygosity thresholds were left as defaults (3 and 0.01 respectively).

After running SISRS, note path to the SISRS loci (aligned contigs) output directory:

* SISRS_Run/aligned_contigs

***
# 01. Filter By Taxa
Scripts are located in the [01_FilterByTaxa](https://github.com/LMBiancani/PlacentalPolytomy/tree/main/01_FilterByTaxa) folder.

* Removes excessively incomplete sequences (sequence must be less than 33% Ns)
* Removes aligned loci with incomplete clade sampling (locus must include sequences for at least 18 of the 37 taxa, approx. 50%)
* Removes aligned loci with incomplete clade sampling (locus must include a sequence from all 4 taxon groups included in the `groups.csv` file)
* The required taxon-to-group correspondence table, `groups.csv`, is a csv file with the following format: \
```Group,Taxa```\
```group1,taxonName1```\
```group1,taxonName2```\
```group2,taxonName3```\
```group2,taxonName4```\

## 01.00 slurm submission script: `01.00_FilterByTaxa.sh`
* runs the following python script: `filter_SISRS_output.py`
* requires the followin input file: `groups.csv`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 01_FilterByTaxa/01.00_FilterByTaxa.sh
```

### Python script: `filter_SISRS_output.py`
*  run by previous shell script: `01.00_FilterByTaxa.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat 01_FilterByTaxa/filter_SISRS_output.py
```

### Taxon-to-group table: `groups.csv`
* path to this input file specified in previous shell script: `01.00_FilterByTaxa.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat 01_FilterByTaxa/groups.csv
```

# 02. Assess Phylogenetic Signal: IQ-TREE
Scripts are located in the [02_IQtree](https://github.com/LMBiancani/PlacentalPolytomy/tree/main/02_IQtree) folder.

## 02.00 slurm submission script: `02.00_iqtree_prep.sh`
* sets up the folders and lists of files to process
* creates output folders for iqtree array scripts
* generates batch fasta files for aligned loci
* generates the array details that need to be added to subsequent array submission scripts
* Note: array details generated by this analysis: `#SBATCH --array=[1-28]%28`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 02_IQtree/02.00_iqtree_prep.sh
```

## 02.01 slurm submission script: `02.01_iqtree_array.sh`
* runs the following R scripts: `trimTrees.R`, `getSCF.R`
* generates CSV filed (one for each batch fasta file produced by `02.00_iqtree_prep.sh`) in the `output/02_iqtree_assessment/02.01_compare_hypotheses/likelihood` directory. CSV files contain likelihood scores for each locus and each hypothesis. Example: `LnLs_1.csv`\
```SISRS_contig-1000033.fasta, -8588.512```\
```SISRS_contig-1000033.fasta, -8589.722```\
```SISRS_contig-1000033.fasta, -8589.617```\
```SISRS_contig-10005000015.fasta, -839.789```\
```SISRS_contig-10005000015.fasta, -839.917```\
```SISRS_contig-10005000015.fasta, -839.959```\
* generates CSV files (one fore each contig) in the `output/02_iqtree_assessment/02.01_compare_hypotheses/scf` directory. CSV files contain a header and 3 lines of SCF values (one for each hypothesesis). Example: `scf_SISRS_contig-1000000007.fasta`\
```ID,sCF,sCF_N,sDF1,sDF1_N,sDF2,sDF2_N,sN,debug```\
```36,13.95,0.68,34.2,1.66,34.85,1.81,4.158,2```\
```36,31.14,1.51,34.54,1.77,14.53,0.66,3.934,2```\
```36,36.26,1.76,30.54,1.45,12.4,0.57,3.778,2```\
* individual likelihood and scf csv files will be combined by `02.01b_iqtree_collect_output.sh`
* Note: update the header with the array details generated by `02.00_iqtree_prep.sh`
* Note: update paths to specified input files (described below)
* The following input files are required and described below: `Polytomy_Placental_Hypotheses.tree`, `tips_Xenarthra.txt`, `tips_Afrotheria.txt`, `tips_Boreoeutheria.txt`, `tips_Outgroup.txt`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 02_IQtree/02.01_iqtree_array.sh
```

### Topology Constraints
Collapsed trees were created for each alternative hypothesis and used as constraint trees in IQtree

Scaffold hypotheis trees:
```
(((XENARTHRA,BOREOEUTHERIA),AFROTHERIA),METATHERIA);
(((AFROTHERIA,XENARTHRA),BOREOEUTHERIA),METATHERIA);
(((AFROTHERIA,BOREOEUTHERIA),XENARTHRA),METATHERIA);
```
Individual collapsed nodes:

METATHERIA:
```
(Didelphis_virginiana,Thylacinus_cynocephalus,Sarcophilus_harrisii,Vombatus_ursinus,Phascolarctos_cinereus,Phalanger_gymnotis,Gymnobelideus_leadbeateri,Pseudochirops_corinnae,Wallabia_bicolor,Potorous_gilbertii)
```
AFROTHERIA:
```
(Procavia_capensis,Loxodonta_africana,Hydrodamalis_gigas,Trichechus_manatus_latirostris,Orycteropus_afer,Elephantulus_edwardii,Amblysomus_hottentotus_longiceps,Chrysochloris_asiatica,Microgale_talazaci,Echinops_telfairi)
```
BOREOEUTHERIA:
```
(Condylura_cristata,Ceratotherium_simum,Manis_javanica,Odobenus_rosmarus_divergens,Hippopotamus_amphibius,Pteropus_vampyrus,Galeopterus_variegatus,Pan_troglodytes,Rattus_norvegicus,Tupaia_tana)
```
XENARTHRA:
```
(Mylodon_darwinii,Choloepus_didactylus,Myrmecophaga_tridactyla,Tamandua_tetradactyla,Dasypus_novemcinctus,Tolypeutes_matacus,Chaetophractus_vellerosus)
```
#### Alternative Hypothesis Trees: `Polytomy_Placental_Hypotheses.tree`
* text file containing alternative hypotheses trees (in Newick format)
* specify path to this input file in previous shell script: `02.01_iqtree_array.sh`
* Keep note of tree order in file:
1. Afrotheria out (Boreoeutheria and Xenarthra are sisters)
2. Boreoeutheria out (Afrotheria and Xenarthra are sisters)
3. Xenarthra out (Afrotheria and Boreoeutheria are sisters)
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat 02_IQtree/hypothesis_trees/Polytomy_Placental_Hypotheses.tree
```

#### Specify a list of focal taxa for each hypothesis tree:
* For placental root topology question: determine which 2 out of 3 groups are sisters for each hypothesis and select one of these 2 sister groups.

##### Focal taxa for tree 1: `tips_Xenarthra.txt`
* Tree 1 = Afrotheria Out, focal taxa = Xenarthra
* specify path to this input file in previous shell script: `02.01_iqtree_array.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat 02_IQtree/hypothesis_trees/tips_Xenarthra.txt
```

##### Focal taxa for tree 2: `tips_Afrotheria.txt`
* Tree 2 = Boreoeutheria Out, focal taxa = Afrotheria
* specify path to this input file in previous shell script: `02.01_iqtree_array.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat 02_IQtree/hypothesis_trees/tips_Afrotheria.txt
```

##### Focal taxa for tree 3: `tips_Boreoeutheria.txt`
* Tree 3 = Xenarthra Out, focal taxa = Boreoeutheria
* specify path to this input file in previous shell script: `02.01_iqtree_array.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat 02_IQtree/hypothesis_trees/tips_Boreoeutheria.txt
```

##### Outgroup taxa list: `tips_Outgroup.txt`
* specify path to this input file in previous shell script: `02.01_iqtree_array.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_text"}
cat 02_IQtree/hypothesis_trees/tips_Outgroup.txt
```

### R script: `trimTrees.R`
* run by the following shell scripts: `02.01_iqtree_array.sh` `02.02_iqtree_array_concat.sh`
* trim taxa off of the main tree(s) if not included in taxon composition of a particular alignment
```{bash, comment=NA, echo=FALSE, class.output="bg_Rcode"}
cat 02_IQtree/trimTrees.R
```

### R script: `getSCF.R`
* run by previous shell script: `02.01_iqtree_array.sh`
* extract sCF values for the branch of interest from IQ-TREE output
```{bash, comment=NA, echo=FALSE, class.output="bg_Rcode"}
cat 02_IQtree/getSCF.R
```

### 02.01b slurm submission script: `02.01b_iqtree_collect_output.sh`
* collects individual fit assessment data
* concatenates likelihood CSV files into combined file: `output/02_iqtree_assessment/02.01_compare_hypotheses/combined_dLnLs.csv` containing likelihood scores for every locus and each hypothesis. Format:\
```SISRS_contig-1000033.fasta, -8588.512```\
```SISRS_contig-1000033.fasta, -8589.722```\
```SISRS_contig-1000033.fasta, -8589.617```\
```SISRS_contig-10005000015.fasta, -839.789```\
```SISRS_contig-10005000015.fasta, -839.917```\
```SISRS_contig-10005000015.fasta, -839.959```\
```...```\
* generates combined scf CSV file: `output/02_iqtree_assessment/02.01_compare_hypotheses/combined_scf.csv`. Combined CSV file contains a line for each locus with the locus name and an sCF value for each hypothesis. Format:\
```SISRS_contig-1000000007.fasta,13.95,31.14,36.26```\
```SISRS_contig-100000000.fasta,15.73,26.46,36.85```\
```SISRS_contig-1000000020.fasta,0.5,2.47,1.13```\
```...```\

```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 02_IQtree/02.01b_iqtree_collect_output.sh
```

## 02.02 slurm submission script: `02.02_iqtree_array_concat.sh`
* runs AMAS to infer concatenated trees and iqtree to calculate likelihoods
* runs the following R scripts: `trimTrees.R`
* Note: update paths to specified input files
* Note: update the header with the array details generated by `02.00_iqtree_prep.sh`

```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 02_IQtree/02.02_iqtree_array_concat.sh
```

### 02.02b slurm submission script: `02.02b_iqtree_collect_phyloinference_LnL.sh`
* collects concatenation fit assessment data
* moves output files from `02.02_iqtree_array_concat.sh` to `02.02_concat_trees/concat_array_files/` sub directory
* generates a CSV file `combined_iqtree_dLnLs_concat.csv` (in `output/02_iqtree_assessment/02.02_concat_trees/`) where each row corresponds to a partition (i.e. locus), and each column contains the cumulative site log-likelihood for that partition under the three different tree hypotheses. Format:\
```locusA,-12345.6,-12347.2,-12340.1```\
```locusB,-10023.4,-10025.9,-10020.8```\
```CHECK OUTPUT FORMAT```\

```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 02_IQtree/02.02b_iqtree_collect_phyloinference_LnL.sh
```

## 02.03 slurm submission script: `02.03_iqtree_array_gtree.sh`
* runs iqtree to infer gene trees
* Note: update the header with the array details generated by `02.00_iqtree_prep.sh`
* individual gene trees will be collected by `02.03b_iqtree_collect_gtrees.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 02_IQtree/02.03_iqtree_array_gtree.sh
```

### 02.03b slurm submission script: `02.03b_iqtree_collect_gtrees.sh`
* collects all individual loci (gene tree) names into `gtrees.txt`
* combines all individual gene trees into `gtrees.tre`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 02_IQtree/02.03b_iqtree_collect_gtrees.sh
```



## need updating:

* iqtree_collect_phyloinference_LnL.sh - slurm script to submit: collect concatenation fit assessment data



# 03. Annotate Loci
Scripts are located in the [03_Annotation](https://github.com/LMBiancani/PlacentalPolytomy/tree/main/03_Annotation) directory.

* Aligns SISRS loci of a particular taxon to a reference genome (ideally, of the same taxon).

## 03.00 slurm submission script: `03.00_DownloadReference.sh`
* script to retreive a reference taxon for the loci
* downloads reference genome (FASTA) and annotation (GFF) for Pan troglodytes (chimpanzee) [GCF_002880755.1](https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_002880755.1/)
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 03_Annotation/03.00_DownloadReference.sh
```

## 03.01 slurm submission script: `03.01_Annotation.sh`
* A python script is used to filter the output and convert it to BED. Overlapping hits of similar scores as well as very disjunct alignments are discarded. The BED file is then sorted and intersected with the GFF file for the reference sequence.
* Another python script is then used to processess the intersected BED file to produce the final output,
either counts of different annotation types, or length proportion of each annotations type. The following types of annotations are recorded: pseudogene, CDS, UTR, intron, lnc_RNA, other (any other type), unannotated (or intergenic).
* runs the following python scripts: `annotation_getTaxContigs.py`, `annotation_blast_parser.py`, `annotation_bed2table.py`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 03_Annotation/03.01_Annotation.sh
```

### Python script: `annotation_getTaxContigs.py`
* run by previous shell script: `03.01_Annotation.sh`
* extracts SISRS loci of a particular taxon into a single file for use with BLAST
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat 03_Annotation/annotation_getTaxContigs.py
```

### Python script: `annotation_blast_parser.py`
* BLAST to BED python script run by the previous shell script: `03.01_Annotation.sh`
* adjust parameters on lines 10-12 as needed
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat 03_Annotation/annotation_blast_parser.py
```

### Python script: `annotation_bed2table.py`
* BED to table python script run by the previous shell script: `03.01_Annotation.sh`
* two options are provided (to be specified in shell script: `03.01_Annotation.sh`)
1. count the number of each feature per locus (`c`), for ex. 1 CDS, 2 introns, etc.
2. compute proportion of length of each feature type per locus (`l`), for ex. 0.2 CDS, 0.8 introns
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat 03_Annotation/annotation_bed2table.py
```

# 04. Feature Assessment
Scripts are located in the [04_FeatureAssessment](https://github.com/LMBiancani/PlacentalPolytomy/tree/main/04_FeatureAssessment) folder. Scripts were adapted from [AlexandraWalling/PML/3_feature_assessment](https://github.com/alexandrawalling/PML/3_feature_assessment).
* Ensure MAFFT, AMAS, IQ-TREE2, FastSP, and HyPhy are installed, update paths/modules in the submission scripts as needed.
* **Note:** AMAS takes file names as command line arguments with a limit on how long the line can be. In order to analyze over a hundred thousand files `04.00_subset.sh` will split loci into subsets.

## 04.00 slurm submission script: `04.00_subset.sh`
* copies alignment files from a source directory into multiple subdirectories, each containing a specified number of files.
* This script splits the 109,799 loci into 20 subsets of 5500 loci (max).
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.00_subset.sh
```

## 04.01 slurm submission script: `04.01_prep.sh`
* Creates loci groups for alignment
* Creates output subdirectories 
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.01_prep.sh
```

## 04.02 slurm submission script: `04.02_mafft_array.sh`
* Uses MAFFT to re-align loci
* Alignments and re-alignments are later compared
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.02_mafft_array.sh
```

## 04.03 slurm submission script: `04.03_amas.sh`
* Runs AMAS to assess loci features
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.03_amas.sh
```

## 04.04 slurm submission script: `04.04_FastSP.sh`
* Runs FastSP to assess loci features
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.04_FastSP.sh
```

## 04.05 slurm submission script: `04.05_iqtree_concat_array.sh`
* Runs AMAS to produce concatenated alignments (`concatenated.fasta`) and partition files (`partitions.txt`) for each subset of 5500 loci in `$out/subset_xx/iqtree_concatttree`
* Runs IQTree to infer a concatenated tree for each loci subset
* A concatenated alignment and tree for ALL is needed
but AMAS will not take that many files as input so the subset alignments will be concatenated together by `04.05b_iqtree_concat_allLoci.sh`

```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.05_iqtree_concat_array.sh
```

### 04.05b slurm submission script: `04.05b_iqtree_concat_allLoci.sh`
* For each subset, copies (and renames to include subset #) concatenated alignments and partition files to `all_loci` directory
* Runs AMAS to concatenate subset alignments
* Creates partitions file for all concatenated loci
* Runs IQTree to infer a concatenated tree for all loci
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.05b_iqtree_concat_allLoci.sh
```

## 04.06 slurm submission script: `04.06_run_iqtree_genetree_array.sh`
* This is a parent script that will submit a separate array job for each subset
* The array jobs will Run IQTree to infer gene trees for each locus
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.06_run_iqtree_genetree_array.sh
```

## 04.07 slurm submission script: `04.07_prune_rescale_species_tree.sh`
* Prunes species tree to only include taxa present in gene tree and rescales
* runs the following R script: `prune_tree.R`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.07_prune_rescale_species_tree.sh
```

### R script: `prune_tree.R`
* R script run by the previous shell script: `04.07_prune_rescale_species_tree.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat 04_FeatureAssessment/scripts/prune_tree.R
```

## 04.08 slurm submission script: `04.08_rate_assessment_array.sh`
* This is a parent script that will submit a separate array job for each subset.
* Note: requires a HyPhy batch script (LEISR.bf): A likelihood estimation script for evolutionary rates. Location of batch script: `PlacentalPolytomy/04_FeatureAssessment/scripts/HyPhy/2.5.33-gompi-2020b/share/hyphy/TemplateBatchFiles/LEISR.bf`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.08_rate_assessment_array.sh
```

## 04.09 slurm submission script: `04.09_run_astral.sh`
* Runs Astral to infer a coalescent tree for each subset of gene trees
* runs the following R script: `collapse_by.R`
* Note: A coalescent tree for ALL gene trees is needed and will be inferred by `04.09b_run_astral_allLoci.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.09_run_astral.sh
```

### 04.09b slurm submission script: `04.09b_run_astral_allLoci.sh`
* Runs Astral to infer a coalescent tree for all gene trees
* runs the following R script: `collapse_by.R`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.09b_run_astral_allLoci.sh
```

### R script: `collapse_by.R`
* R script run by the previous shell scripts: `04.09_run_astral.sh` and `04.09b_run_astral_allLoci.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat 04_FeatureAssessment/scripts/collapse_by.R
```

## 04.10 slurm submission script: `04.10_assess_properties.sh`
* combines assessment output files for all loci
* runs the following R script: `assess_gene_properties.R`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat 04_FeatureAssessment/04.10_assess_properties.sh
```

### R script: `assess_gene_properties.R`
* R script run by the previous shell scripts: `04.10_assess_properties.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_python"}
cat 04_FeatureAssessment/scripts/assess_gene_properties.R
```


# 05. Figures

%%%%%%%%%%%%%% left off here %%%%%%%%%%%%%%%%

# AMAS
* Runs AMAS to assess locus features
* AMAS is run in batches by the driver script (`run_amas.py`).

#### Download AMAS
* note path to `AMAS.py` and add path to slurm script `run_amas.sh`
```
git clone https://github.com/marekborowiec/AMAS/
```
#### Slurm submission script: `run_amas.sh`
* runs the following python script: `run_amas.py`
```{bash, comment=NA, echo=FALSE, class.output="bg_slurm"}
cat amas/run_amas.sh
```
#### Python script: `run_amas.py`
* python script run by shell script: `run_amas.sh`
```{bash, comment=NA, echo=FALSE, class.output="bg_Rcode"}
```

***