\documentclass[../main.tex]{subfiles}
\begin{document}

Unresolved phylogenetic relationships in the Tree of Life can negatively impact species classification \citep{Som2015-et}, conservation \citep{Swenson2009-om}, and our understanding of traits \citep{}, among other topics of interest to biologists.
At the same time, even modern phylogenetic analyses frequently result in some tree nodes being effectively unresolved, either due to poor statistical support or to conflicting outcomes after reanalysis by alternative methods \citep{Rodriguez-Ezpeleta2007-vn, Smith2015-ae, Esselstyn2017-uc,Chakrabarty2017-ie}. This is due to the prevalence of non-phylogenetic signal and noise when trying to resolve such contentious relationships.

Sources of phylogenetic error can be broadly divided into stochastic and systematic factors \citep{Kapli2021-dd}. Stochastic error results from prevailing random noise in an insufficient amount of data \citep{Kapli2021-dd} and is typically well-managed in large contemporary phylogenomic-scale datasets \citep{Philippe2011-fv}. On the other hand, systematic error is due to consistent differences in the data, either between taxa (sequences), between loci, or both \citep{Kapli2021-dd, Philippe2011-fv}. As systematic error is independent of the size of the dataset, it is generally more tricky to address \citep{Philippe2011-fv, Jeffroy2006-th}.

Systematic error can be caused by an array of analytical factors including data acquisition and processing artefacts \citep{Simion2017-jz, Philippe2011-fv}, incorrect homology inference \citep{Fernandez2020-gb}, alignment errors \citep{Ranwez2020-le}, missing data \citep{Wiens2006-cm}, and even software errors \citep{Simion2020-ex}. While most analyses strive to only include orthologous sequences in analyses \citep{Boussau2020-bk, Fernandez2020-gb}, the presence of paralogs is arguably one of the biggest methodological issues in modern-day phylogenomic datasets \citep{Simion2020-ex, Brown2017-hn}. While in-paralog noise is minor and poses no significant issues for inference \citep{Yan2022-up}, divergent out-paralogs (which we refer to as ``deep paralogs" later in the text) can create significant artefactual relationships \citep{Springer2018-av, Springer2018-mk, Gatesy2017-zn}. Additional systematic impacts on the phylogenetic analyses can come from an array of biological properties of analyzed taxa, including a high level of Incomplete Lineage Sorting (ILS) \citep{Degnan2009-hn}, differences in rate of molecular evolution, biases in base composition, or variation in strength of selection on the substitutions in the sequence.

One common approach to reduce systematic error is locus filtering or subsampling \citep{Edwards2016-gr, Gilbert2018-im, Smith2022-pj}. In this approach loci with suboptimal properties indicative of elevated noise levels are excluded from the phylogenetic inference, thus maximizing signal to noise ratio \citep{Edwards2016-gr,Chen2015-tr,Simmons2016-pe, Molloy2018-tr}. Several criteria have been used to rank loci in order to filter out least optimal ones, including missing data \citep{Mongiardino_Koch2021-ai, Evangelista2021-cr, Kocot2017-wt, Molloy2018-tr, Brown2017-hn, Herrando-Moraira2018-vw}, taxon occupancy \citep{Mongiardino_Koch2021-ai, Borowiec2015-mr, Evangelista2021-cr, Gernandt2018-qn, Herrando-Moraira2018-vw,Lemmon2009-tj, Young2016-mc, Dietrich2017-ix}, proportion of variable or parsimony informative sites and substitution rate \citep{Mongiardino_Koch2021-ai, Borowiec2015-mr, Whelan2015-gk, Herrando-Moraira2018-vw, Gernandt2018-qn}, alignment length \citep{Mongiardino_Koch2021-ai, Evangelista2021-cr, Brown2017-hn, Gernandt2018-qn}, base composition heterogeneity \citep{Mongiardino_Koch2021-ai, Evangelista2021-cr,Kocot2017-wt,Whelan2015-gk}, saturation level \citep{Mongiardino_Koch2021-ai, Borowiec2015-mr, Kocot2017-wt, Herrando-Moraira2018-vw}, average bootstrap support of a gene tree \citep{Mongiardino_Koch2021-ai, Borowiec2015-mr, Herrando-Moraira2018-vw}, and phylogenetic informativeness \citep{Mongiardino_Koch2021-ai, Herrando-Moraira2018-vw, Townsend2007-zi}. Most approaches have focused on independent treatment of the filtering criteria by applying them sequentially, although subsampling based on a combined score of multiple different criteria has been tried as well \citep{Herrando-Moraira2018-vw}.

The effectiveness of subsampling as a mitigation approach for phylogenetic noise is still a highly debated topic \citep{Simmons2016-pe, Molloy2018-tr}. Some authors have argued that removing data from analyses can do more harm than good \citep{Chan2020-bz}. However, sequencing or initial bioinformatics is a form of subsampling that occurs by default in all datasets since no molecular datasets from diverse species can include entire genome sequences in analyses. In most cases a reduced molecular dataset is obtained and used by selection of particular regions either in advance or after genome sequencing and processing.
For example, single copy protein coding orthologs \citep{Simao2015-eh, Zhang2019-ok}, small conserved regions of the genome \citep{Lemmon2012-or, Schwartz2015-wg, Literman2021-yc}, or ultraconserved elements \citep{McCormack2012-tz, Zhang2019-ok, Chakrabarty2017-ie, Esselstyn2017-uc} are frequently used in phylogenetics.
Fortunately, with modern-day genome-scale data, even subsampling retains extremely large datasets with extensive information.
%Even when reference-free methods are used on whole genome sequencing data, only a subset of data is included in final analyses .

A caveat of subsampling is that it is only effective  as a mitigation
approach for particular types of non-phylogenetic signal or noise.
%when systematic noise is heterogeneous across the dataset. Homogeneously and randomly distributed confounding
For example, loci that experienced incomplete lineage sorting (ILS) can produce accurate gene trees that do not match the species tree.
%where could not be addressed by subsampling, and
The effect of these loci is best tackled by  analytical approaches such as using coalescent-aware phylogenetic software \citep{Zhang2018-sb,Chifman2014-gs}.
Thus, despite considerable recent emphasis on the impacts of ILS on the conflicting relationships \citep{Edwards2009-uk, Rannala2020-zb}, this type of discordance is beyond the scope of the present paper.

\rs{Add Smith et all parrott paper in here} \ak{This section specifically talks about the MK paper, their conclusion, so that then I can explain how we build on it. Smith et al probably can be cited somewhere upstream. I already put it in few paragraphs above as per your previous comment}
Recent work has suggested that subsampling by simultaneously analyzing several confounding factors, and ranking data based on individual or their combined effects in order to filter out noisy data, can improve species tree estimation \citep{Mongiardino_Koch2021-ai}. %\citet{Mongiardino_Koch2021-ai} assessed several properties of and performed a principal component analysis (PCA) to determine relationships between properties.
%Data from multiple empirical amino acid sequencing datasets was subsampled based on  \citep{Mongiardino_Koch2021-ai}.
%In the results, \citet{Mongiardino_Koch2021-ai} determined that
Subsampling based on high statistical support of gene trees led to the best outcomes, while ranking based on missing data was ambiguous with respect to improving phylogenetic signal ratio \citep{Mongiardino_Koch2021-ai}. Loci with average substitution rates also produced the best phylogenetic inference, with rates that were too high or too low producing suboptimal results \citep{Mongiardino_Koch2021-ai}.
However, it is unclear whether the subsampling-induced changes in the inferred species tree were improvements, since only empirical datasets, without known true trees, were considered. At the same time, the best performing subsampling criterion was found to be RF-distance to the inferred species tree, a conclusion with some degree of circularity as RF distance is unknown unless the true tree is already known \citep{Mongiardino_Koch2021-ai}.
%This analysis also focused on amino acid (protein-coding) datasets, primarily in invertebrates, with certain property assessments not readily applicable to nucleotide-based datasets.
%Additionally, while the analysis of \citet{Mongiardino_Koch2021-ai} did show similarity of certain data features related to the filtering and phylogenetic inference results, it remains largely unknown to what extent and how linearly different features interact with each other when performing subsampling.

In this study we examined phylogenetic data filtering by leveraging simulated datasets with controlled locus properties and known true species histories. Using machine learning in the form of a random forest regressor allowed us to predict the level of phylogenetic utility in any locus, given its properties, as well as to better understand interactions between different factors contributing to the noise.
Our trained model accurately predicted the phylogenetic utility of the test datasets.
Reanalyzing subsets of data with high and low predicted phylogenetic utility based on the trained model provided information on the impacts of the subsetting on the topologies
in empirical datasets \citep{Fong2012-hy, McGowen2020-ru, Wickett2014-as, Liu2017-wo}.
%especially for difficult nodes identified in the original studies and branch length estimation.
We show that multiple factors can contribute to non-phylogenetic signal, and these factors can interact. The effectiveness of model-based filtering on empirical data varied among datasets, but the subsets of loci with the most non-phylogenetic signal tended to have a different inferred topology and poor statistical support
suggesting these data can impact phylogenetic estimation, and removing them using this approach can benefit phylogenetic estimation.

%\ak{whatever the term we use for noise, I see this as including everything, both lack of information and conflicting information etc. I think it's nice to have several words (non-phylogenetic signal, noise, error, etc) so that text can have less repetitions of the same term over and over. Break down of the chosen term into various types is done in the next paragraph. I couldnt find an ideal way to talk about this, partly also because some things can classified multiple ways. Feel free to restructure / rewrite this and the next sections in a different way.}

% Despite dramatic increases in the amount of data used for phylogenetic reconstruction, many relationships among organisms remain unresolved or controversial.
% Reasons for this include differences in phylogenetic signal among loci (among other things such as methodological differences, modeling, and others).
% Differences in phylogenetic signal can manifest as strong conflict among genes favoring different topologies and lead to low confidence in placement of contentious lineages.
% Fortunately, with genome-scale data comes the potential to evaluate the quality of the signal in loci \rs{cite Literman; cite Knyshov} and therefore support for particular hypotheses.

% The mechanisms of the contribution of non-phylogenetic signal to phylogenetic conflicts is still a highly debated topic \citep{Schrempf2020-ge}.
% Sources of non-phylogenetic signal are broadly divided into stochastic and systematic factors \citep{Kapli2021-dd}. Stochastic error results from prevailing random noise in insufficient amount of data and is typically well-managed in modern day phylogenomic-scale datasets (Philippe et al 2017). On the other hand systematic error can be present regardless of the size of the dataset (Philippe et al 2017) and caused by an array of methodological and biological factors\citep{Boussau2020-bk}. Methodological factors include data acquisition and processing artefacts \citep{Simion2017-jz, Philippe2011-fv}, incorrect homology inference \citep{Fernandez2020-gb}, alignment errors \citep{Ranwez2020-le}, and even software errors \citep{Simion2020-ex}. Biological factors contributing to gene conflict can be classified into homoplasy, when incorrect relationships are inferred due to convergence in the sequences among distantly related organisms, and hemiplasy, when incorrect relationships are inferred due to incomplete lineage sorting (ILS) or gene flow \citep{Hibbins2020-av, Mendes2019-zy}.

% Data filtering is a common strategy in which loci with suboptimal properties indicative of elevated noise levels are excluded from the phylogenetic inference thus maximizing signal to noise ratio (TBD Graybeal, 1994; Meyer et al. 2011; Chen et al. 2015; Edwards 2016; Simmons et al. 2016; Molloy and Warnow 2018). We note that most if not all molecular datasets do not include entire genome sequences in analysis but rather use specific subsets of the data, for example, only single copy protein coding orthologs, or only UCE, or other preselected markers. However, before the data collection it is hard to know the amount of phylogenetic signal, so it is during the data analysis when the additional filtering occurs, which we refer to as subsampling.

% The effectiveness of subsampling as a mitigation approach for phylogenetic noise is still a highly debated topic (i.e., Simmons et al., 2006, Molloy and Warnow 2018). Filtering is expected to be effective when systematic noise is heterogeneous across the dataset. Homogeneously and randomly distributed confounding factors, such as incomplete lineage sorting (ILS), could not be addressed by subsampling, and instead are best tackled by changing analytical approaches. For example, in the case of ILS, doing coalescent-aware phylogenetics leads to more accurate inference results (cite Astral, SVDQ, StarBEAST here). Thus, despite much emphasis has been put on the impacts of ILS on the conflicting relationships \citep{Edwards2009-uk, Rannala2020-zb}, this type of discordance is beyond the scope of the present paper.

% On the other hand, when there is a variation in levels of non-phylogenetic signal across loci, subsampling would be a good strategy to employ.
% \ak{small section on paralogy here} While most analyses strive to only include orthologous sequences into analyses \citep{Boussau2020-bk, Fernandez2020-gb}, the presence of paralogs is arguably one of the biggest methodological issues in modern day phylogenomic datasets \citep{Simion2020-ex, Brown2017-hn}. It has been shown that in-paralog noise is minor and poses no significant issues for the inference \ak{ref}, however, divergent out-paralogs can create significant artefactual relationships \ak{ref}.
% \ak{small section on missing data or maybe make it all one section}

% \ak{prefinal section - highlight how these properties, mentioned above, were mostly considered independently when investigating their impacts on phylogeny inference. Specifics about papers trying to make sense of factors separately and in combination}

% In this study build on previous empirical dataset-based studies in context of phylogenetic signal filtering by leveraging simulated datasets with controlled locus properties and known true species histories. We utilize machine learning in the form of a random forest regressor to predict the level of phylogenetic utility in any locus given its properties as well to better understand interactions between different factors contributing to the noise. We trained the model on simulated data, ranked the loci by predicted utility values, and reanalyzed subsets of data with best and worst phylogenetic utility to gauge the improvements in reconstructing the true tree. We further used the model to predict utility of loci in several diverse empirical datasets, filter the worst loci, and re-infer the phylogenies. We measure impacts of the subsetting on the overall topologies, difficult nodes identified in the original studies, as well as branch length distribution.
% \rs{add brief results statements and indicate potential impacts on the discipline}
% We show that multiple factors can contribute to non-phylogenetic signal, and these factors can additionally interact. Our trained model accurately predicted phylogenetic utility of the test datasets. Effectiveness of model-based filtering on empirical data varied between datasets, but the subsets of loci with most non-phylogenetic signal tended to have a different inferred topology and poor statistical support.



\end{document}
